tools/loader/ImageClassifier.cpp

```diff
@@ -457,34 +492,87 @@ int main(int argc, char **argv) {
       }
 
       inputImagePH = inputOutputPair.first;
-      SMT = inputOutputPair.second;
+      outputPH = inputOutputPair.second;
+      SMT = bindings->get(outputPH);
     }
     assert(inputImagePH && SMT && "Input and output must be valid.");
-    GLOW_ASSERT(inputImagePH->dims() == inputImageData.dims() &&
+    GLOW_ASSERT(inputImagePH->dims() == inputImageData->dims() &&
                 "New input shape does not match the compiled function.");
 
     // Convert the raw input to fp16. This must be done every time we get new
     // image data.
     if (convertInAndOutToFp16) {
-      inputImageData.convertToType(ElemKind::Float16Ty);
+      inputImageData->convertToType(ElemKind::Float16Ty);
     }
 
     // About to run inference, so update the input image Placeholder's backing
     // Tensor with inputImageData.
-    updateInputPlaceholders(bindings, {inputImagePH}, {&inputImageData});
+    updateInputPlaceholders(*bindings, {inputImagePH}, {inputImageData.get()});
 
     // Perform the inference execution, updating SMT.
-    auto batchSize = inputImageData.dims()[0];
-    loader.runInference(bindings, batchSize);
-
-    // Print the top-k results from the output Softmax tensor.
-    numErrors += processAndPrintResults(SMT, inputImageBatchFilenames);
+    auto batchSize = inputImageData->dims()[0];
+    if (iterationsOpt) {
+      unsigned requestCount = 1;
+      if (miniBatch) {
+        requestCount = iterationsOpt / miniBatch;
+      }
+      // Setup list of inference requests to be run.
+      for (unsigned i = 0; i < requestCount; i++) {
+        auto newImage = llvm::make_unique<Tensor>(inputImageData->getType());
+        auto newContext = llvm::make_unique<ExecutionContext>();
+        auto ph = newContext->getPlaceholderBindings();
+        ph->allocate(inputImagePH);
```

Comment:
Am I reading it correctly that this loop is going to allocate `requestCount` batches of images ahead of time?  If I wanted to run 10,000 iterations of resnet50 (not unrealistic for the Habana chip) would I need to allocate 6 GB of memory?  That doesn't seem desirable for a benchmark...

Reply:
That is correct, the thought here was we want each image to be a different location, we could reuse the same context N many times but wouldn't we see better results than expected due to caching?

Reply:
We'd have a similar caching effect in the production system I think since that's more or less how we allocate there. I think maybe you can abstract that somewhat by using a TensorPool with a fixed size, with a condition variable that you can block waiting for old requests to finish.